api/client/charms/localcharmputters.go

```diff
@@ -0,0 +1,149 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package charms
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"strconv"
+
+	"github.com/juju/charm/v12"
+	"github.com/juju/errors"
+	"gopkg.in/httprequest.v1"
+
+	"github.com/juju/juju/api"
+	"github.com/juju/juju/rpc/params"
+)
+
+type CharmPutter interface {
+	PutCharm(ctx context.Context, modelUUID, charmRef, curl string, body io.Reader) (string, error)
+}
+
+type HTTPPutter struct {
+	httpClient *httprequest.Client
+}
+
+func NewHTTPPutter(apiConn api.Connection) (CharmPutter, error) {
+	// The returned httpClient sets the base url to /model/<uuid> if it can.
+	apiHTTPClient, err := apiConn.HTTPClient()
+	if err != nil {
+		return nil, errors.Annotate(err, "cannot retrieve http client from the api connection")
+	}
+	return &HTTPPutter{
+		httpClient: apiHTTPClient,
+	}, nil
+}
+
+func (h *HTTPPutter) PutCharm(ctx context.Context, _, _, curlStr string, body io.Reader) (string, error) {
+	curl, err := charm.ParseURL(curlStr)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	args := url.Values{}
+	args.Add("series", curl.Series)
+	args.Add("schema", curl.Schema)
+	args.Add("revision", strconv.Itoa(curl.Revision))
+	apiURI := url.URL{Path: "/charms", RawQuery: args.Encode()}
+
+	var resp params.CharmsResponse
+	req, err := http.NewRequest("POST", apiURI.String(), body)
+	if err != nil {
+		return "", errors.Annotate(err, "cannot create upload request")
+	}
+	req.Header.Set("Content-Type", "application/zip")
+
+	if err := h.httpClient.Do(ctx, req, &resp); err != nil {
+		return "", errors.Trace(err)
+	}
+	return resp.CharmURL, nil
+}
+
+type S3Putter struct {
+	httpClient *httprequest.Client
+}
+
+func NewS3Putter(apiConn api.Connection) (CharmPutter, error) {
+	apiHTTPClient, err := apiConn.RootHTTPClient()
+	if err != nil {
+		return nil, errors.Annotate(err, "cannot retrieve http client from the api connection")
+	}
+	return &S3Putter{
+		httpClient: apiHTTPClient,
+	}, nil
+}
+
+func (h *S3Putter) PutCharm(ctx context.Context, modelUUID, charmRef, curl string, body io.Reader) (string, error) {
+	apiURI := url.URL{Path: fmt.Sprintf("/model-%s/charms/%s", modelUUID, charmRef)}
+
+	resp := &http.Response{}
+	req, err := http.NewRequest("PUT", apiURI.String(), body)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	req.Header.Set("Content-Type", "application/zip")
+	req.Header.Set("Juju-Curl", curl)
+
+	if err := h.httpClient.Do(ctx, req, &resp); err != nil {
+		return "", errors.Trace(err)
+	}
+
+	return resp.Header.Get("Juju-Curl"), nil
+}
+
+type FallbackPutter struct {
```

Comment:
Is there a different name we can use for this? Trying to think of one as well.

Reply:
Admittedly there is room for improvement here. Happy to change it if anyone thinks of anything. I'm of the impression FallbackPutter will do, especially now we don't export it outside of this module